### PR TITLE
Add sort by "required" and then name

### PR DESCRIPTION
**Pull request by sjauld**
So I was thinking something like this for #42. Might be nice to have this as a command line option but I'll need to do a bit more digging on how to get that working.

### CODE REVIEW

**Comment by sjauld** [comment]
Command line flag added!

**Comment by jufemaiz** [comment]
Hey guys, wondering if there's a possibility to do anything with this?

**Comment by metmajer** [comment]
Hi @sjauld.

Thank you for your interest in this project!

With the release of [v0.4.0](../releases/v0.4.0), the project has undergoing major refactoring to facilitate future community contributions like yours. Likewise, in the absence of tests, maintaining a high-quality codebase was close to impossible.

That said, the feature you've implemented is a great addition to `terraform-docs`. Would you be interested in updating your branch with the latest changes to master?

**Comment by sjauld** [comment]
No problem! I have refactored and added a test for the sorting.

**Comment by metmajer** [comment]
@sjauld This looks like a great first-time contribution. I'll review your commits shortly!

File: internal/pkg/doc/doc_test.go [hunk 1]
```diff
@@ -243,6 +243,88 @@ func TestInputsFromVariablesTf(t *testing.T) {
 	assert.Equal(t, expected, actual)
 }
 
+func TestInputsSortedByRequired(t *testing.T) {
```
Comment (metmajer) [inline internal/pkg/doc/doc_test.go]: Great you've added a test for the `doc.SortInputsByRequired` method. May I ask you to create the missing test for `doc.SortInputsByName`?

File: internal/pkg/doc/input.go [hunk 2]
```diff
@@ -53,3 +53,32 @@ func (a inputsSortedByName) Less(i, j int) bool {
...
+func (a inputsSortedByRequired) Swap(i, j int) {
+	a[i], a[j] = a[j], a[i]
+}
+
+func (a inputsSortedByRequired) Less(i, j int) bool {
+	switch {
+	// i required, j not: i gets priority
+	case !a[i].HasDefault() && a[j].HasDefault():
+		return true
+		// j required, i not: i does not get priority
```
Comment (metmajer) [inline internal/pkg/doc/input.go]: The comment is not proper indented. Are you using `go fmt`?

File: internal/pkg/print/print.go [hunk 3]
```diff
@@ -15,6 +15,8 @@ const (
 	WithAggregateTypeDefaults
 	// WithRequired prints if inputs are required
 	WithRequired
+	// WithSortByRequired prints with required fields sorted to the top
```
Comment (metmajer) [inline internal/pkg/print/print.go]: We could be more precise here. Example: "WithSortByRequired prints inputs sorted by "required", then by name.

File: internal/pkg/print/print.go [hunk 4]
```diff
@@ -15,6 +15,8 @@ const (
...
+	WithSortByRequired
 	// WithSorting prints sorted inputs and outputs
 	WithSorting
```
Comment (metmajer) [inline internal/pkg/print/print.go]: In light of the newly added WithSortByRequired, you could rename this to WithSortByName.

File: main.go [hunk 5]
```diff
@@ -17,7 +17,7 @@ var version = "dev"
 
 const usage = `
   Usage:
-    terraform-docs [--no-required] [--no-sort] [--with-aggregate-type-defaults] [json | markdown | md] <path>...
+    terraform-docs [--no-required] [--no-sort] [--sort-by-required] [--with-aggregate-type-defaults] [json | markdown | md] <path>...
```
Comment (metmajer) [inline main.go]: Please rename to `--with-sort-inputs-by-required`.

File: main.go [hunk 6]
```diff
@@ -44,6 +44,7 @@ const usage = `
 	-h, --help                       show help information
 	--no-required                    omit "Required" column when generating markdown
 	--no-sort                        omit sorted rendering of inputs and ouputs
+	--sort-by-required               sort the required inputs to the top of the output table
```
Comment (metmajer) [inline main.go]: Suggest to adapting name and description as described above.

**Comment by metmajer** [comment]
@sjauld I've seen that the `AUTHORS` file is not in the changeset. If you run `make all`, the file will be updated with your name as a contributor.

File: internal/pkg/doc/input.go (hunk 2 above)
Comment (sjauld) [reply internal/pkg/doc/input.go]: Hmmm, gofmt didn't pick that one up, I guess because it's inline with the item above!

**Comment by sjauld** [comment]
Thanks for the feedback - should be good to go now :man_dancing:

**Comment by metmajer** [comment]
Great work, @sjauld 👍

I see the following things missing:
- Add tests `TestPrintWithSortInputsByRequired` (underneath `TestPrintWithSortByName`) in…
    - `print/json`
    - `print/markdown`
    - `print/pretty`
- Update README.md to show the new option.

Please complete so that we can close this PR. Thanks!

File: internal/pkg/print/print.go [hunk 7]
```diff
@@ -15,8 +15,10 @@ const (
 	WithAggregateTypeDefaults
 	// WithRequired prints if inputs are required
 	WithRequired
-	// WithSorting prints sorted inputs and outputs
-	WithSorting
+	// WithSortInputsByRequired prints inputs with required fields first and then by name
```
Comment (metmajer) [inline internal/pkg/print/print.go]: Ask asked before, please change this description to something that's better understandable. Preferrably "WithSortInputsByRequired sorts inputs by name and prints required inputs first". Please also use this where you define `--sort-inputs-by-required` in `main.tf`s and in the documentation in `main.tf` and `README.md`.

**Comment by metmajer** [comment]
Hey @sjauld, I consider your contribution very valuable. Are you planning on providing the missing things?

**Comment by sjauld** [comment]
Hi @metmajer,

Yes - sorry about the delay! I didn't have any spare time last week but I'll try to finish up in the next couple of days. 

 Cheers,

**Comment by sjauld** [comment]
Ok, all ready to go!

**Comment by metmajer** [comment]
Congratulations @sjauld, this is an excellent contribution!

**Pull request merged**